packages/node_modules/@webex/plugin-meetings/src/reachability/index.js

```diff
@@ -0,0 +1,283 @@
+
+import ReachabilityRequest from './request';
+
+/**
+ * @class Reachability
+ */
+export default class Reachability extends StatelessSparkPlugin {
+  namespace = MEETINGS;
+
+  constructor(attrs, options) {
+    super({}, options);
+    this.reachabilityRequest = new ReachabilityRequest({}, options);
+    this.gatherReachability();
+  }
+
+  setup = () => {
+    this.processedClusterResults = [];
+    this.peerConnectionMap = {};
+  }
+
+  /**
+   * fetches reachability data
+   * @returns {object} reachability data
+   */
+  gatherReachability = () => {
+    this.setup();
+    // Reset reachability result to empty to ensure no stale data
+    try {
+      window.localStorage.setItem(REACHABILITY.localStorage, JSON.stringify({}));
+
+      // Get clusters from Orpheus
+      return this.reachabilityRequest.getClusters()
+        .then((clusters) =>
+          // Perform Reachability Check
+          this.performReachabilityCheck(clusters)
+            .then((response) => {
+              window.localStorage.setItem(REACHABILITY.localStorage, JSON.stringify(response));
+              return response;
+            }))
+        .catch((error) => {
+          console.info(`Error in calling getClusters() :${error}`);
+          return {};
+        });
+    }
+    catch (error) {
+      console.log('There was an error while accessing LocalStorage', error);
+      return {};
+    }
+  };
+
+  /**
+   * fetches reachability data
+   * @param {object} clusterList
+   * @returns {object} reachability check results
+   */
+  performReachabilityCheck = (clusterList) => {
+    if (!clusterList || !clusterList.length) {
+      return Promise.resolve({});
+    }
+
+    let localSDPList = [];
+
+    return new Promise(((resolve) => {
+      this.getLocalSDPForClusters(clusterList)
+        .then((localSDPData) => {
+          if (!localSDPData || !Object.keys(localSDPData).length || !localSDPData.localSDPList || !localSDPData.peerConnection) {
+            // TODO: handle the error condition properly and try retry
+            console.info('Local SDP is empty or has missing elements..returning.');
+            resolve({});
+          }
+
+          this.peerConnectionMap = localSDPData.peerConnection;
+          ({localSDPList} = localSDPData);
+          this.setRemoteOfferForEachCluster(localSDPList)
+            .then((response) => {
+              resolve(response);
+            });
+        })
+        .catch((error) => {
+          console.error(`Error in getLocalSDPForClusters :${error}`);
+          resolve({});
+        });
+    }));
+  };
+
+  /**
+   * sets remote offer for each clusters
+   * @param {object} localSDPList local SDP generated
+   * @returns {object} reachability check results
+   */
+  setRemoteOfferForEachCluster = (localSDPList) => {
+    const iceResult = {};
+    return new Promise((resolve) => {
+      this.reachabilityRequest.remoteSDPForClusters(localSDPList)
```

Comment:
Since you don't have a return here any outer promise isn't guaranteed to wait for the result of this.

Reply:
It does in a different way . it waits for the ice candidate and resolves it when the length is equal . Its kind of tricky.